<!--
This starts up the pose_graph creator node.  
Start this together with move_base_stage, for example, and drive the robot around in rviz.
It will create a pose graph and publish it to the visualization_marker topic.
   
-->

<launch>

  
  <arg name="slam" default="true"/>
  <arg name="fixed_frame" default="odom"/>
  <arg name="db_location" default="$(find laser_slam)/data"/>

  <!-- KL: this node was already commented...
  <node pkg="laser_slam" type="top_graph.py" name="top_graph"
        args="laser_slam_node" output="screen" />
  -->

  <remap from="global_map" to="map" if="$(arg slam)"/>
  <remap from="scan" to="base_scan" />

  <node name="mongo" type="wrapper.py" pkg="mongodb" output="screen">
    <param name="overwrite" value="true"/>
    <param name="database_path" value="$(arg db_location)"/>
  </node>


  <node pkg="laser_slam" type="laser_slam_node" name="laser_slam_node" output="screen">

    <param name="optimization_frame" value="map" if="$(arg slam)"/>
    <param name="optimization_frame" value="graph_optimization" unless="$(arg slam)"/>
    <remap from="initialpose" to="stage_init_pose"/>

    <param name="optimization_algorithm" value="spa"/>
    <param name="add_new_nodes" value="true"/>
    
    <param name="position_displacement_threshold" value="0.5" />
    <param name="angle_displacement_threshold" value="0.5" />
    
    <param name="fixed_frame" value="$(arg fixed_frame)"/>
    <param name="base_frame" value="base_footprint"/>
    <param name="laser_frame" value="base_laser_link"/>
    
    <param name="scan_time_tolerance" value="0.1"/>

    <param name="grid_resolution" value="0.025" />
    <param name="visualization_rate" value="0.3" />

    <param name="odom_noise" value="0.0" />

    <param name="grid_pub_rate" value="0.3"/>
    <param name="scan_match_proportion" value="0.2"/>

    <param name="cleanup_grid" value="false"/>
    
  </node>
  

  <node pkg="laser_slam" type="cloud_snapshotter" name="cloud_snapshotter" output="screen"/>

  <node pkg="laser_slam" type="scan_match_constraint_generator" name="constraint_generator" output="screen"/>

  <node pkg="laser_slam" type="scan_match_localizer" name="scan_match_localizer" output="screen"/>


</launch>
